### PR TITLE
Falcon40b prefill attention optimizations

### PR DESCRIPTION
**Pull request by johanna-rock-tt**
- Use partially sliced and sharded attention sequence
- Use better softmax implementation that reduces data transfer for attention mask

[Multichip pipeline passed: https://github.com/tenstorrent/tt-metal/actions/runs/8797794117](https://github.com/tenstorrent/tt-metal/actions/runs/8833259521)

### CODE REVIEW

File: models/demos/falcon40b/tt/falcon_attention.py [hunk 1]
```diff
@@ -382,69 +380,72 @@ def fwd_prefill(
...
+        if num_slices > 1:
+            # Partially sliced and sharded attention
+            attn_output = []  # this is the output we write to. Initiate as empty tensors
+            for i in range(len(query_layer)):
+                attn_output.append(
+                    torch2tt_tensor(
+                        torch.zeros([1, self.num_heads_per_device, q_len, self.head_dim]),
+                        self.devices[i],
+                        tt_memory_config=self.model_config["DRAM_MEMCFG"],
+                        tt_dtype=self.model_config["POST_SOFTMAX_MM_OUTPUT_DTYPE"],
```
Comment (s-jovic) [inline models/demos/falcon40b/tt/falcon_attention.py:395]: Can we avoid this? In the decoder we will sum up attention output with residual and deallocate it - maybe we can skip deallocation and just pass it in to the next layer since the values inside don't matter?

I am not sure how much time we spend to push this tensor to device, if it is negligible then it probably doesn't make sense since it might not be intuitive passing in a tensor whose value is not used anywhere.

File: models/demos/falcon40b/tt/falcon_attention.py [hunk 2]
```diff
@@ -469,8 +469,51 @@ def fwd_prefill(
...
+            )
+
+        # Softmax
+        for i in range(len(attn_weights)):
+            attn_weights[i] = tt_lib.operations.primary.transformers.scale_causal_mask_hw_dims_softmax_in_place(
+                attn_weights[i],
+                self.scalar,
+                attn_mask_slices[i],
+                program_config=self.model_config["SOFTMAX_PROGCFG"],
+                # is_causal_mask=True,
```
Comment (s-jovic) [inline models/demos/falcon40b/tt/falcon_attention.py:497]: Is this a leftover comment or we might need it for debug?

File: models/demos/falcon40b/tt/model_config.py [hunk 3]
```diff
@@ -821,6 +836,9 @@ def get_prefill_model_config(model_config_str, input_shape, num_devices):
     model_config["KV_CACHE_MEMCFG"] = DRAM_MEMCFG
     model_config["KV_CACHE_DTYPE"] = BFP8_DTYPE
 
+    # model_config["DENSE_H_TO_4H_MM_WEIGHTS_DTYPE"] = BFP4_DTYPE
+    # model_config["DENSE_4H_TO_H_MM_WEIGHTS_DTYPE"] = BFP4_DTYPE
+
```
Comment (s-jovic) [inline models/demos/falcon40b/tt/model_config.py:841]: Do we need this?

File: models/demos/falcon40b/tt/model_config.py [hunk 4]
```diff
@@ -838,6 +856,39 @@ def get_prefill_model_config(model_config_str, input_shape, num_devices):
...
+        attn_shard_spec = ttl.tensor.CoreRangeSet(
+            {
+                ttl.tensor.CoreRange(
+                    ttl.tensor.CoreCoord(0, 0),
+                    ttl.tensor.CoreCoord(7, 3),
+                ),
+            }
+        )
+    else:
+        attetnion_mm_grid_size = (8, 8)
```
Comment (s-jovic) [inline models/demos/falcon40b/tt/model_config.py:883]: nit: typo

**Review by s-jovic — APPROVED** [review]

File: models/demos/falcon40b/tt/falcon_attention.py (hunk 1 above)
Comment (pavlepopovic) [reply models/demos/falcon40b/tt/falcon_attention.py:395]: Agree, could also move the initial allocation to model_preprocessing?

**Review by pavlepopovic — APPROVED** [review]

File: models/demos/falcon40b/tt/falcon_model.py [hunk 5]
```diff
@@ -165,14 +165,8 @@ def model_preprocessing(self, llm_mode, input_ids, kv_cache_len, num_input_token
...
 
-            attention_mask_heads_dim = (
-                self.config.num_attention_heads
-                if self.model_config["ATTN_MASK_MEMCFG"].is_sharded()
-                else len(self.devices)
-            )
-
             attention_mask_bool_chunks = torch.chunk(
-                (attention_mask_bool * -100000).expand(-1, attention_mask_heads_dim, -1, -1),
+                (attention_mask_bool * -100000).expand(-1, len(self.devices), -1, -1),
```
[the remark this answers is not in the document]
Comment (johanna-rock-tt) [reply models/demos/falcon40b/tt/falcon_model.py:169]: Did you test this? Currently it's bfp8. The values are 0 or -100000 in that case. Would any type of mask containing 0s and -100000s result in the same bfp4_b and bfp8_b values? Like: a block of 16x16 0s / -100000s / and a mixture of both?

File: models/demos/falcon40b/tt/falcon_attention.py (hunk 1 above)
Comment (johanna-rock-tt) [reply models/demos/falcon40b/tt/falcon_attention.py:395]: Good point. The problem is, that the size of this depends on the sequence length, it can be different from inference to inference. So I don't think we can pre-allocate it. In theory we could reuse it for all layers, that's what @s-jovic meant, right? But before we sum it up we also AllGather over all devices, where we allocate a different memory area anyway, right? So...I guess we could allocate it the first time and then reuse, and not free it up after the AllGather? Do you think that makes sense in this case? I have no idea how long allocation/deallocation takes, I could measure and then see if it makes sense to avoid it?

Comment (s-jovic) [reply models/demos/falcon40b/tt/falcon_attention.py:395]: `model_preprocessing` already depends on the seq len, right? We call it for each user anyhow

File: models/demos/falcon40b/tt/falcon_model.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (pavlepopovic) [reply models/demos/falcon40b/tt/falcon_model.py:169]: Would be worth trying trying out. I guess if it's working on BFP8 it should also work on BFP4

File: models/demos/falcon40b/tt/falcon_attention.py (hunk 1 above)
Comment (s-jovic) [reply models/demos/falcon40b/tt/falcon_attention.py:395]: Sure, makes sense

File: models/demos/falcon40b/tt/falcon_model.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (johanna-rock-tt) [reply models/demos/falcon40b/tt/falcon_model.py:169]: With bfp4 I get good pcc and no hang for S=128, but a hang (1st attempt) and an ARC scratch error (2nd attempt) for S=2048. I'll use bfp8 for now but create an issue for this since it seems to be functionally correct in the S=128 case.

**Comment by johanna-rock-tt** [comment]
Seeing a hang with S=2048 and 12 layers; will investigate

**Review by arakhmati — APPROVED** [review]
